docs: add design pages for runtime recording feature (#118)
tier4/CARET_doc: docs/design/runtime_processing/runtime_recording.md

# Runtime recording

> As explained in [Tracepoint section](../trace_points/index.md), CARET records meta-information at initialization and reduces tracepoint data to as much as possible at runtime.

Reviewer: > reduces tracepoint data to as much as possible
Maybe we don't need "to" in this case.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

This allows for a low overhead at runtime, but recording including meta-information requires a LTTng session before running the application.

In actual operation, though, it is often desirable to start recording during executions.

Therefore, CARET stores meta-information in memory,

> and record the meta-information at the start of recording.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> This session explains the details of runtime recording feature.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

See also:

- [Software architecture | caret_trace](../software_architecture/caret_trace.md)

## Overview

> The runtime recording is a feature that stores initialization information in memory and records it to trace data after measurement starts, enabling recording even during execution.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> This function consists of the following states

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

- WAIT state
  - Obtain information on running applications and store trace data in memory.
- PREPARE state
  - Record stored trace data as LTTng tracepoints (delayed recording).
- RECORD state
  - Record runtime trace data as LTTng trace points (synchronous recording).

These states are controlled by nodes (trace node) running in an independent thread.
trace nodes run in each process.

<prettier-ignore-start>
!!! Notice
    The thread on which a trace node runs is created via function hooking.
    This thread is created even if the Application is not implemented by rclcpp.
    This allows runtime recording as well for nodes implemented in rclpy. (Note for it cannot record runtime trace data correctly.)
    Although python has a Global Interpreter Lock (GIL) mechanism,
    it runs as an asynchronous thread that does not even get a GIL.
<prettier-ignore-end>

> Typical use cases are as follows.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

```bash
# Run a node at Terminal 0 first.
ros2 run pkg node
```

```bash
# Execute "record" command with Termial 1 after node startup.
ros2 caret record
```

State transition is shown below.

```plantuml
concise "Terminal 0" as User0
concise "Terminal 1" as User1
concise "trace node state" as Hook
concise "LTTng session" as Lttng

@0
User0 is "Idle"
User1 is "Idle"
Hook is "Idle"
Lttng is "Idle"

@2
User0 is "ros2 run pkg node"
Hook is WAIT : \nStore initialization \ntracepoints into memory\n.
User0 -> Hook

@6
User1 is "ros2 caret record"
Hook is PREPARE : Record stored \ninitialization tracepoints
User1 -> Hook : Start recording \n[ /caret/start_record ]\n\n
User1 -> Lttng
Lttng is "Active"

@10
Hook is RECORD : Record initialization \nand runtime tracepoints
Hook -> User1 : Notify state transition \n[ /caret/status ]\n

@15
User1 is "Idle"
Hook is WAIT
User1 -> Hook : End recording \n[ /caret/end_record ]\n\n
User1 -> Lttng
Lttng is "Destroyed"
```

> A detailed sequence diagram is written in [Sequence](#sequence).

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> Trace nodes uses topic communication for control and notification.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> In addition, to maintain compatibility of usage,

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> in cases where sessions are started in advance, they can record correctly.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

```plantuml
concise "Terminal 0" as User0
concise "trace node state" as Hook
concise "Terminal 1" as User2
concise "LTTng" as Lttng

@0
User0 is ""
User2 is ""
Hook is ""
Lttng is "idle"

@3
User2 is "ros2 caret trace"
User2 -> Lttng : Start session\n
Lttng is "Active"

@5
User0 is "launch"
Hook is RECORD : Record initialization and runtime tracepoints
User0 -> Hook

@15
User2 is ""
User2 -> Lttng : Destroy session\n
Lttng is "Destroyed"

```

> Note that Initialization trace data is recorded for each LTTng session.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> The main state transitions are shown below.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

```plantuml
[*] --> RECORD: An active lttng session exists
[*] --> WAIT: No active lttng session exists
WAIT : Stores initialization trace data in memory
PREPARE : Record initialization trace data with LTTng
WAIT --> PREPARE :Start recording

RECORD --> WAIT : End recording
PREPARE --> RECORD : Finished recording stored initialization trace data
RECORD : Record initialization and runtime trace data with LTTng
```

> Detailed state transitions are written in [Status](#status).

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

## Topic

> Runtime recording uses the following topic communication.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

| topic name            | message type | role                                          |
| --------------------- | ------------ | --------------------------------------------- |
| `/caret/start_record` | Start.msg    | Start recording. Transition to PREPARE state. |
| `/caret/end_record`   | End.msg      | End recording. Transition to WAIT state.      |
| `/caret/status`       | Status.msg   | Sync current recording state.                 |

### Start.msg

```cpp
uint32 recording_frequency 100
string ignore_nodes  # reserved
string ignore_topics # reserved
string select_nodes  # reserved
string select_topics # reserved
```

> The recording_frequency is the frequency at which each process records initialization trace data in the PREPARE state.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> The higher the frequency, the shorter the time required to complete the PREPARE state, but the greater the risk of tracer discarded.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> `ignore_nodes` `ignore_topics` `select_nodes` `select_topics` are unimplemented features.

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

They are reserved fields for setting [tracepoint filtering](./tracepoint_filtering.md) at the start of the measurement from CLI.

<prettier-ignore-start>
!!!Info

>     As another method of recording_frequency,

Reviewer 2: ```suggestion
    As an alternative of `recording_frequency`,
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/5

>     initialization-related trace points can be recorded in [blocking mode](https://lttng.org/blog/2017/11/22/lttng-ust-blocking-mode/) to prevent tracer discarded.

Reviewer 2: ```suggestion
    [blocking mode](https://lttng.org/blog/2017/11/22/lttng-ust-blocking-mode/) might be a reasonable choice to avoid tracer discarded during storing meta-data into a disk.
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/5

>     This method reduces the occurrence of data loss, but this time we decided to improve it on the caret side to reduce implementation changes.

Reviewer 2: I'd like to check how you decided why you rejected `blocking mode`

<prettier-ignore-end>

### Status.msg

```cpp
int8 UNINITIALIZED=0
int8 WAIT=1
int8 PREPARE=2
int8 RECORD=3

string caret_node_name
int8 status
string[] node_names # reserved
int64 pid # reserved
```

> `trace_node` is a trace node name.

Reviewer 2: ```suggestion
A trace node name is given to `caret_node_name` field.
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> `status` is the WAIT, PREPARE, or RECORD status.

Reviewer 2: ```suggestion
`status` field has one of states; WAIT, PREPARE, or RECORD state.
```

> `node_names` is an unimplemented feature.

Reviewer 2: ```suggestion
`node_names` field is unused for an unimplemented feature.
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

It is a reserved field to represent a list of node names managed by the trace node.

> `pid` is an unimplemented feature.

Reviewer 2: ```suggestion
`pid` field is also an unimplemented feature.
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

It is a reserved field to represent the process ID.

### End.msg

```cpp
(Empty)
```

The End topic is for notification, so its contents are empty.

> ## Status

Reviewer 2: ```suggestion
## State definition
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

> A detailed state transition description for each state is shown below.

Reviewer 2: ```suggestion
A detailed state transition is described below.
```

Reviewer 2: fixed at https://github.com/hsgwa/CARET_doc/pull/6/files

| item                               | description                                                                                                                                                  |
| ---------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Transition conditions for entering | - Receive messages from `/caret/start_record` topic.                                                                                                         |
| Transition conditions for exiting  | - Receive messages from `/caret/end_record` topic. <br> - Finish recording stored initialization trace data.                                                 |
| Initialization trace data          | - Record as LTTng tracepoint (synchronous recording). <br> - Record stored data as LTTng tracepoint at fixed frequency from trace nodes (delayed recording). |
| Runtime trace data                 | - Discard to prevent discarding initialization trace data.                                                                                                   |

> The recording frequency of Stored initialization trace data can be changed with `recording_frequency` in Start.msg.

Reviewer 2: I'd like to edit this sentence with holding a meeting with @hsgwa -san.
